### PR TITLE
adapt core/autocrypt to PGPy in acpgpy/autocrypt

### PR DESCRIPTION
**Pull request by juga0**
This PR can:

- be left as it is (after fixing bugs and adapting doc), we would have 2 implementation versions
- adapt core/autocrypt to add acpgpy/autocrypt code and choose version via command line
- remove core/autocrypt and leave this one (after fixing bugs and adapting doc)

I squashed commits so all this version is in only one.

Some tests are failing in python3 (probably just str/unicode) and some need to be reviewed because they might not have sense with the PGPy version (marked with TODO or FIXME).

To see the differences between the two implementations: `diff core/autocrypt acpgpy/autocrypt`.

@hpk42, @dkg, opinions?

### CODE REVIEW

**Comment by hpk42** [comment]
Thanks for the PR! I can't use the github review interface for this because everything is added anew (a result from the copy-operation you did) and thus i have no diffs to comment on here.  It seems i can locally just copy your acpgpy/* files over to the autocrypt/ dir and look at the diffs but then i can't comment here on it, either.  So could you do this copy operation yourself and remove acpgpy?  This way we can discuss the diff here ... and is it right that the core of your work is currently in acpgpy/crypto.py? 

The gpg backend needs to stay for the time being because I, and i think others, need integration into their existing gpg state -- i am fine to consider to make "bingpg" only an option but I won't through out this code and  rely solely on pgpy in the next couple of months. 

I am preparing and leaving for travel in a few hours until sunday and will try to then get back to the PR on monday -- please update it as mentioned before.  I do not consider it crucial to get pgpy merged before easterhegg btw -- [the session there](https://eh17.easterhegg.eu/Fahrplan/events/8499.html) i sketched as an overview of email encryption, some autocrypt approaches, probably abstract UI demos, and then a hands-on session where people try to get it to work if they have a suitable setting (most notably a local postfix instance which configures the autocrypt milter).  I want to get some releases out which people can pip-install which also needs to happen and be tested early next week.  And trying/refining the UI demos is obviously a priority if we want to use them in the session ...

That said, let's get pgpy into the core -- i'd like to have it and am happy to try it as default backend.  It's great that you went through the effort of persevering in making it work ... it'd be very nice to have a self-contained python package which can do autocrypt and does not depend on random versions of GPG.

**Comment by juga0** [comment]
> So could you do this copy operation yourself and remove acpgpy?

I have done so and `push -f` so this PR is now based on that.

> and is it right that the core of your work is currently in acpgpy/crypto.py?

yes, now called autocrypt/crypto.py.

> The gpg backend needs to stay for the time being

that is fine also for me.
 
> I do not consider it crucial to get pgpy merged before easterhegg 

it would be nice though, as people could then contribute to that version.

> I want to get some releases out which people can pip-install which also needs to happen and be tested early next week.

yeah, that would be nice too, it would be even better if that pip-install would include the pgpy version, but i know we have limited time.

> the UI demos is obviously a priority if we want to use them in the session

agree.

> That said, let's get pgpy into the core -- i'd like to have it and am happy to try it as default backend. It's great that you went through the effort of persevering in making it work ... it'd be very nice to have a self-contained python package which can do autocrypt and does not depend on random versions of GPG.

great, thank *you*.

File: .travis.yml [hunk 1]
```diff
@@ -33,7 +33,13 @@ script:
   - gpg2 --version
   - echo $PATH
   - which gpg2
-  - tox -e $TESTENV --recreate -- --no-test-cache --with-gpg2 tests/
+  # - tox -e $TESTENV --recreate -- --no-test-cache --with-gpg2 tests/
+  - cd core
```
Comment (hpk42) [inline .travis.yml]: why another "cd core"?  There is already one above and the travis build fails because of this double "cd core".

File: core/.coveragerc [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+source =
+    acpgpy
+   .tox/*/lib/python*/site-packages/acpgpy
+   .tox/*/lib/site-packages/acpgpy
+   .tox/pypy*/site-packages/acpgpy
+
+[report]
+omit =
+ */python?.?/*
+ *__init__*
```
Comment (hpk42) [inline core/.coveragerc]: the "acpgpy" references look wrong now.  Also i'd prefer to discuss/add coverage and other testing in a separate PR because this PR is already quite big ...

File: core/AUTHORS [hunk 3]
```diff
@@ -0,0 +1,6 @@
+Holger krekel <holger at merlinux dot eu>
+juga <juga at riseup dot net>
+azul <azul at riseup dot net>
+Daniel Kahn Gillmor <dkg at fifthhorseman dot net>
+Vincent Breitmoser <look at my dot amazin dot horse>
+Oliver Wiese <oliver dot wiese at fu-berlin dot de>
```
Comment (hpk42) [inline core/AUTHORS]: did you check that all these authors have significant (say >1% contributions to the code base?)

File: core/CHANGELOG.rst [hunk 4]
```diff
@@ -7,3 +7,7 @@
   but you can glimpse from the tests how to use
   it.
 
+0.3.2
+-------
+
+- acpgy: core based version using PGPy library
```
Comment (hpk42) [inline core/CHANGELOG.rst]: see above, reference to acpgpy don't make sense anymore.

File: core/autocrypt/__init__.py [hunk 5]
```diff
@@ -1,2 +1,47 @@
...
+    included in all copies or substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+    EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+    IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+    CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+    TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+    SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+"""
```
Comment (hpk42) [inline core/autocrypt/__init__.py]: again, i'd like to have all the additions to metadata and here the license addition in a separate PR.

As to license i usually use "MIT license" as reference and then there is a license file in the root.

File: core/autocrypt/account.py [hunk 6]
```diff
@@ -38,11 +41,11 @@ def _dict(self):
...
-    # def _reload(self):
-    #    try:
-    #        self._property_cache.pop("_dict")
-    #    except AttributeError:
-    #        pass
+    def _reload(self):
+        try:
+            self._property_cache.pop("_dict")
+        except AttributeError:
+            pass
```
Comment (hpk42) [inline core/autocrypt/account.py]: why did you need to uncomment this?

File: core/autocrypt/account.py [hunk 7]
```diff
@@ -295,18 +285,33 @@ def process_incoming(self, msg, delivto=None):
...
-                    keyhandle = ident.bingpg.import_keydata(keydata)
+                    logger.debug('type(d[key]) %s', type(d["key"]))
+                    logger.debug('type(keydata) %s', type(keydata))
+                    keyhandle = ident.crypto.import_keydata(keydata)
+                    logger.debug('imported key with keyhandle %s', keyhandle)
+                    pgpykey = ident.crypto.get_publickey_from_kh(keyhandle)
+                    ident.crypto.export_key(pgpykey)
+                    logger.debug('exported key')
+                    ident.crypto.publicpgpykeys.append(pgpykey)
                     d["*keyhandle"] = keyhandle
```
Comment (hpk42) [inline core/autocrypt/account.py]: please remove most if not all logger.debug() calls -- it's nice during debugging but not as a permanent addition.  Note that you can also do ``import pdb ; pdb.set_trace()`` and use interactive commands to get at variable values so it's not always neccessary to print all values all the time.

File: core/autocrypt/mime.py [hunk 8]
```diff
@@ -88,6 +91,7 @@ def parse_ac_headervalue(value):
         if name == "key":
             value = "".join(value.split())
         result_dict[name] = value
+        logger.debug('len(result_dict) %s', len(result_dict))
     return result_dict
```
Comment (hpk42) [inline core/autocrypt/mime.py]: also here (and in other places) please remove logger.debug calls -- apart from pdb.set_trace() you can also just use "print()" commands which will be shown for failing tests in a special section.

File: core/requirements.txt [hunk 9]
```diff
@@ -0,0 +1,3 @@
+click==6.7
+-e git+https://github.com/SecurityInnovation/PGPy.git@release/0.4.1#egg=PGPy-0.4.1
```
Comment (hpk42) [inline core/requirements.txt]: please use a release from pypi, not some git reference -- also what do you need "requirements.txt" for?

File: core/setup.py [hunk 10]
```diff
@@ -30,7 +32,19 @@ def main():
...
             autocrypt=autocrypt.cmdline:autocrypt_main
         ''',
-        install_requires = ["click>=6.0", "six"],
+        install_requires=[
+            "click>=6.0",
+             "six"
+        ],
+        dependency_links=[
+            "https://github.com/SecurityInnovation/PGPy.git@\
+             @release/0.4.1#egg=PGPy-0.4.1"
```
Comment (hpk42) [inline core/setup.py]: this also needs to be a normal install_requires pinned release -- unless this is not a released version of pgpy in which case we need to wait until they release it. (dependency_links does not work in some pip versions and requires extra options for "pip install" ASFAIU and no project i am involved in uses it).

File: core/autocrypt/crypto.py [hunk 11]
```diff
@@ -0,0 +1,574 @@
...
+            return x
+
+    def set(self, val):
+        propcache = self.__dict__.setdefault("_property_cache", {})
+        propcache[f.__name__] = val
+    return property(get, set)
+
+
+class Crypto(object):
+    """ GPG operations for Autocrypt using PGPy. """
```
Comment (hpk42) [inline core/autocrypt/crypto.py]: GPG?

File: core/autocrypt/crypto.py [hunk 12]
```diff
@@ -0,0 +1,574 @@
...
+    # TODO: Write in identity YAML file the keys, so it won't be needed
+    # to import/export in different files.
```
Comment (hpk42) [inline core/autocrypt/crypto.py]: i don't understand this comment ... what do you refer to when you say "identity YAML file"?

File: core/autocrypt/crypto.py [hunk 13]
```diff
@@ -0,0 +1,574 @@
...
+                            usage={KeyFlags.EncryptCommunications,
+                                   KeyFlags.EncryptStorage})
+            logger.debug('subkey added')
+        if protected is True:
+            passphrase = getpass.getpass()
+            skey.protect(
+                             passphrase,
+                             SymmetricKeyAlgorithm.AES256,
+                             HashAlgorithm.SHA256)
+            logger.debug('Key protected')
```
Comment (hpk42) [inline core/autocrypt/crypto.py]: currently py-autocrypt does not support passphrases by design and if we added passphrases we cannot just call into interactive "getpass()" here.  So better remove the support for interactive input here.

File: core/autocrypt/crypto.py [hunk 14]
```diff
@@ -0,0 +1,574 @@
...
+        assert type(pgpykey) == PGPKey
+        if not pgpykey.is_public:
+            pgpykey = pgpykey.pubkey
+        publickeydata = self.get_public_keydata(armor=True,
+                                                pgpykey=pgpykey)
+        pkpath = os.path.join(self.pgpydir, pgpykey.fingerprint.keyid
+                              + '.asc')
+        with open(pkpath, 'w') as fd:
+            fd.write(publickeydata)
+            logger.debug('written %s', pkpath)
```
Comment (hpk42) [inline core/autocrypt/crypto.py]: why does exporting a public key perform a write operation on the file system here? 
Also "export_pkey" is ambigious as it could mean "export_private_key" and only is clearer when one also knows about "export_skey" and figures out the "s" means secret.

File: core/autocrypt/crypto.py [hunk 15]
```diff
@@ -0,0 +1,574 @@
...
+            pgpykey = self.own_pgpykey
+        assert type(pgpykey) == PGPKey
+        assert not pgpykey.is_public
+        secretkeydata = self.get_secret_keydata(armor=True,
+                                                pgpykey=pgpykey)
+        skpath = os.path.join(self.pgpydir, pgpykey.fingerprint.keyid
+                              + '.key')
+        with open(skpath, 'w') as fd:
+            fd.write(secretkeydata)
+            logger.debug('written %s', skpath)
```
Comment (hpk42) [inline core/autocrypt/crypto.py]: why perform a write operation here?

**Comment by hpk42** [comment]
So i have made a number of inline comments and have a few meta comments: 

- please remove logger.debug invocations almost completely, they make it harder to read the code and mostly seem like having been added for development

- you removed bingpg.py and related command line options and handling but it needs to stay.  My original suggestion (a few weeks ago) was to **copy** "bingpg.py" to "pgpy.py" (which you call "crypto.py"), and then make modifications to the other code to call into pgpy/crypto code instead of bingpg.  I offered to then help to properly make pgpy an option.  Your approach rather followed the idea to replace  bingpg related code which now makes it harder for me to make pgpy just an option.  Your approach also makes the PR much bigger than it otherwise needed to be ...

- i am confused by the way how crypto.py handles/stores key information in the file system -- i'd think that only create and import of keys would create a file (maybe using the fingerprint as filename or so) and all other operations would just read from them.  Also i am not sure yet i understand why you needed to add ``config.own_key`` -- maybe it makes sense but why is a keyhandle not enough? 

- lastly your PR contains unrelated changes to metadata (in setup.py, requirements, coverage etc.) which i'd like to keep separate from adding pgpy support.  Again makes the PR much larger.

File: .travis.yml (hunk 1 above)
Comment (juga0) [inline .travis.yml]: ops, mistake

File: core/.coveragerc (hunk 2 above)
Comment (juga0) [inline core/.coveragerc]: yeah, might not have updated all acpgpy correctly.

ok, let's discuss testing in other PR

File: core/AUTHORS (hunk 3 above)
Comment (juga0) [inline core/AUTHORS]: didn't check if >1%, but they all in the logs

File: core/CHANGELOG.rst (hunk 4 above)
Comment (juga0) [inline core/CHANGELOG.rst]: ack

File: core/autocrypt/__init__.py (hunk 5 above)
Comment (juga0) [inline core/autocrypt/__init__.py]: ok

File: core/autocrypt/account.py (hunk 6 above)
Comment (juga0) [inline core/autocrypt/account.py]: probably i didn't need and again mistake cause of merging back to your code

File: core/autocrypt/account.py (hunk 7 above)
Comment (juga0) [inline core/autocrypt/account.py]: i still like to keep them, the user don't see them and new (or old) developers can always know what's going on. Even as a user i always prefer to have verbose output. Could you ignore them atm or is it easier for you to review the PR without them?

File: core/autocrypt/mime.py (hunk 8 above)
Comment (juga0) [inline core/autocrypt/mime.py]: see comment above.

File: core/requirements.txt (hunk 9 above)
Comment (juga0) [inline core/requirements.txt]: i like to give possibilities for installing in different ways. What's wrong with it?.

Comment (juga0) [inline core/requirements.txt]: Regarding the release, 0.4.1 is not in pypi yet.

File: core/setup.py (hunk 10 above)
Comment (juga0) [inline core/setup.py]: I use dependency_links to create debian packages.
Re. the release, same comment as above.

File: core/autocrypt/crypto.py (hunk 11 above)
Comment (juga0) [inline core/autocrypt/crypto.py]: sry, you know well, s/GPG/OpenPGP.
When we go to the comments about the "real code" functionality and not metadata?

File: core/autocrypt/crypto.py (hunk 12 above)
Comment (juga0) [inline core/autocrypt/crypto.py]: Not relevant right now. I'll explain that later. This is a comment...

File: core/autocrypt/crypto.py (hunk 13 above)
Comment (juga0) [inline core/autocrypt/crypto.py]: I made this before adapting it to your code in this repo.
It's just "support" and it's not executed by default, It doesn't hurt either.

File: core/autocrypt/crypto.py (hunk 14 above)
Comment (juga0) [inline core/autocrypt/crypto.py]: I forgot to re-add this comment https://github.com/autocrypt/py-autocrypt/blob/master/core/experiments/gpg_utils.py#L95 there.
Re. pkey, ok, i'll rename it.

File: core/autocrypt/crypto.py (hunk 15 above)
Comment (juga0) [inline core/autocrypt/crypto.py]: Same as above.

**Comment by juga0** [comment]
Regarding to your comments to the comments, metadata and meta-comments:
- Where is the "contributing style guide" to this project?, do we want/need to create it?, collectively?, how are the style rules dictated?.
  After finished reading these comments, it seems to me now that we need it, besides the early stages of this project and our limited time.
- Please don't talk about history/facts that are not reflected in this PR or you don't provide references to them. First, i disagree with them, second, i can provide references and explanations to my disagreement, but that's out of scope in this PR.

Regarding to the comments i was (wrongly) enthusiastic to get:
- Did you run the code?,  and the tests?,  did they work?, isn't PGPy awesome even if still 0.4.1 isn't in pypi?
